src/main.rs

```diff
@@ -0,0 +1,482 @@
+    );
+
+    let graph_chart = Chart::new(datasets)
+        .x_axis(
+            Axis::default()
+                .style(Style::default().fg(Color::White))
+                .bounds(x_bounds),
+        )
+        .y_axis(
+            Axis::default()
+                .style(Style::default().fg(Color::White))
+                .bounds(y_bounds),
+        )
+        .bg(Color::Black);
+
+    frame.render_widget(graph_chart, area);
+}
+
+fn draw_bar_chart<B>(frame: &mut Frame<B>, area: Rect, db: f32, cents: f32)
+where
+    B: Backend,
+{
+    let db_bar = Bar::default()
+        .value((db + 100.0) as u64)
+        .text_value(format!("{:5.01}db", db));
+
+    let cents_bar = Bar::default()
+        .value((cents + 50.0) as u64)
+        .text_value(format!("{:+6.02}c", cents));
+
+    let bar_chart = BarChart::default()
+        .block(Block::default())
+        .direction(Direction::Horizontal)
+        .data(BarGroup::default().bars(&[db_bar, cents_bar]))
+        .max(100);
+
+    frame.render_widget(bar_chart, area);
+}
+
+fn initialize_panic_handler() {
+    let original_hook = std::panic::take_hook();
+    std::panic::set_hook(Box::new(move |panic_info| {
+        crossterm::execute!(std::io::stderr(), crossterm::terminal::LeaveAlternateScreen).unwrap();
+        crossterm::terminal::disable_raw_mode().unwrap();
+        original_hook(panic_info);
+    }));
+}
+
+fn main() -> io::Result<()> {
+    initialize_panic_handler();
+    // setup terminal
+    enable_raw_mode()?;
+    let mut stdout = io::stdout();
+    execute!(stdout, EnterAlternateScreen, EnableMouseCapture)?;
+    let backend = CrosstermBackend::new(stdout);
+    let mut terminal = Terminal::new(backend)?;
+
+    let host = cpal::default_host();
+    let device = host.default_input_device().unwrap();
+    let mut supported_configs_range = device
+        .supported_input_configs()
+        .expect("error while querying configs");
+    let supported_config = supported_configs_range
+        .next()
+        .expect("no supported config?!")
+        .with_max_sample_rate();
+
+    let config = supported_config.config();
+    let (sender, reciever) = std::sync::mpsc::channel::<f32>();
+    let stream = device
+        .build_input_stream(
+            &config,
+            move |data: &[f32], _: &cpal::InputCallbackInfo| {
+                // samples from different channels are received interleaved
+                // mix all the channels into single one by averaging
+                for b in data.chunks_exact(config.channels as usize) {
+                    let val = b.iter().sum::<f32>() / config.channels as f32;
+                    sender.send(val).unwrap();
+                }
+            },
+            move |err| panic!("failed to get microphone input: {}", err),
+            None,
+        )
+        .unwrap();
+    stream.play().unwrap();
+
+    let mut samples = Vec::new();
+    let mut min_amount = BLOCK_SIZE.max(FFT_SIZE);
+
+    let mut state = State::default();
```

Comment:
Consider implementing the `State::with_sample_rate(sample_rate`) method rather than implementing the `default` trait and changing `state`. Also, you could try using this [crate ](https://crates.io/crates/derivative) for a more flexible default configuration.

Reply:
Have you thought about incorporating a GIF demonstration into your project's README? It could serve as an excellent means of showcasing your project.